dask_pytorch/data.py

```diff
@@ -0,0 +1,97 @@
+"""
+Utilities for loading PyTorch data in distributed environments
+"""
+
+
+import tempfile
+from os.path import basename, dirname
+
+from PIL import Image
+from torch.utils.data import Dataset
+
+
+"""
+In the following, we are explicitly avoiding s3fs because it does not behave well with
+multiprocessing (which is commonly used in PyTorch dataloaders
+
+https://github.com/dask/s3fs/issues/369
+"""  # pylint: disable=pointless-string-statement
+
+
+def get_all_files(bucket: str, prefix: str, s3_client=None):
+    """
+    Get list of all files from an s3 bucket matching a certain prefix
+    """
+    import boto3  # pylint: disable=import-outside-toplevel
+
+    if s3_client is None:
+        s3_client = boto3.client("s3")
+    paginator = s3_client.get_paginator("list_objects")
+    all_files = []
+    for page in paginator.paginate(Bucket=bucket, Prefix=prefix):
+        files = [x["Key"] for x in page["Contents"]]
+        all_files.extend(files)
+    return all_files
+
+
+def read_s3_fileobj(bucket, path, fileobj):
+    """
+    read an obj from s3 (in a bucket a a certain path) to a file like object
+    """
+    import boto3  # pylint: disable=import-outside-toplevel
+
+    s3 = boto3.resource("s3")
+    bucket = s3.Bucket(bucket)
+    bucket.download_fileobj(path, fileobj)
+    fileobj.seek(0)
+    return fileobj
+
+
+def load_image_obj(fileobj):
+    """
+    turn a file like object into an image
+    """
+    return Image.open(fileobj).convert("RGB")
+
+
+class BOTOS3ImageFolder(Dataset):
```

Comment:
Could this be called `S3ImageFolder` instead? I think the fact that it happens to use `boto3` is an implementation detail that users shouldn't have to care about. The most important thing, from the perspective of someone importing this library, is "I use this because I have image data living in S3".

Reply:
My point is that nothing about the interface is specific to boto3. It's just a class that lets you manipulate image files in S3, and you don't have to care about what client library happens to be used in the internals.

I'm not proposing that we ever allow people to pick and choose between `boto3` and `s3fs`. I understand that the choice of `boto3` vs. `s3fs` was something interesting that you and Hugo had to deal with while developing this, but it shouldn't matter to someone using this library. They just need to know that this class wraps an S3 bucket full of images and has an API for accessing those images.

In the same way that Dask array isn't called `NumpyDaskArray`, the default client from `elasticsearch` isn't called `RequestsElasticsearch`, etc.

Reply:
Fair! In that case calling it `S3ImageFolder` sounds fine to me.